### ISSUE TITLE
Differenct Results (MNIST) between two versions (TinyDNN)

### ISSUE THREAD

**Issue by l0I0I0l** [tiny-dnn/tiny-dnn#720]
I was running a modified MNIST CNN example.
Find different results.
Old version: **tiny-dnn-master-2016-09-22**
New version: **tiny-dnn-master-2017-05-20**
Both use **AVX** kernels of Tiny DNN.

**Problem is**: I run them several times, all show that **old version** have **much better accuracy**!
Enabling avx/multithreading may have slightly better results but same conclusion!

It seems that both have same constant learning rates and same initializers of weights/biases.
Is it only caused by different numerical **calculation accuracy** or any other reasons?

My modified Example Code in **old version** (tiny-dnn-master-2016-09-22):
```c++
void my_sample1_convnet(const string & data_dir)
{
	//Model Changes:
	// 1. use max-pooling instead of average-pooling
	// 2. use relu instead of tanh
	// 3. use identity after pooling
	// 4. use sigmoid output instead of tanh/softmax
	// 5. use 0 for white pixels (input data)
	//
	const float white = 0.0f; //background
	const float black = 1.0f; //foreground
	const int max_epochs = 20;
	const int minibatch_size = 10;

	network<sequential> nn;
	nn	<< convolutional_layer<relu>(32, 32, 5, 1, 6)
		<< max_pooling_layer<activation::identity>(28, 28, 6, 2)
		<< convolutional_layer<relu>(14, 14, 5, 6, 16)
		<< max_pooling_layer<activation::identity>(10, 10, 16, 2)
		<< convolutional_layer<relu>(5, 5, 5, 16, 120)
		<< fully_connected_layer<sigmoid>(120, 10)
		;
	std::cout << "load models..." << std::endl;
	std::vector<label_t> train_labels, test_labels;
	std::vector<vec_t> train_images, test_images;
	parse_mnist_labels(data_dir + "/train-labels-idx1-ubyte", &train_labels);
	parse_mnist_images(data_dir + "/train-images-idx3-ubyte", &train_images, white, black, 2, 2);
	parse_mnist_labels(data_dir + "/t10k-labels-idx1-ubyte", &test_labels);
	parse_mnist_images(data_dir + "/t10k-images-idx3-ubyte", &test_images, white, black, 2, 2);

	adagrad optimizer;
	optimizer.alpha *= std::sqrt(minibatch_size);
	printf("optimizer.alpha = %f\n", optimizer.alpha);

	progress_display disp(train_images.size());
	timer t;
	int epoch_done = 0;
	auto on_enumerate_epoch = [&](){
		printf("%.2fs elapsed. (%d/%d)\n", t.elapsed(), ++epoch_done, max_epochs);

		tiny_dnn::result res = nn.test(test_images, test_labels);
		std::cout << res.num_success << "/" << res.num_total << std::endl;
		disp.restart(train_images.size());
		t.restart();
	};
	auto on_enumerate_minibatch = [&](){ disp += minibatch_size; };
	nn.train<mse>(optimizer, train_images, train_labels, minibatch_size, max_epochs, on_enumerate_minibatch, on_enumerate_epoch);
	std::cout << "end training." << std::endl;
}
/* traning log (old version, simplified):
Threads:   <Multi>     <Single>
epoch 1:   9877/10000  9876/10000
epoch 2:   9901/10000  9904/10000
epoch 3:   9912/10000  9908/10000
epoch 4:   9918/10000  9916/10000
epoch 5:   9921/10000  9917/10000
epoch 6:   9922/10000  9916/10000
epoch 7:   9919/10000  9914/10000
epoch 8:   9917/10000  9918/10000
epoch 9:   9921/10000  9917/10000
epoch 10:  9923/10000  9915/10000
epoch 11:  9924/10000  9915/10000
epoch 12:  9925/10000  9915/10000
epoch 13:  9924/10000  9914/10000
epoch 14:  9927/10000  9914/10000
epoch 15:  9924/10000  9914/10000
epoch 16:  9923/10000  9916/10000
epoch 17:  9923/10000  9917/10000
epoch 18:  9924/10000  9917/10000
epoch 19:  9924/10000  9917/10000
epoch 20:  9925/10000  9918/10000
*/

```

My modifed Example Code in **new version** (tiny-dnn-master-2017-05-20):
```c++
void my_sample1_convnet(const string & data_dir)
{
	//Model Changes:
	// 1. use max-pooling instead of average-pooling
	// 2. use relu instead of tanh
	// 3. use identity after pooling
	// 4. use sigmoid output instead of tanh/softmax
	// 5. use 0 for white pixels (input data)
	//
	const float white = 0.0f; //background
	const float black = 1.0f; //foreground
	const int max_epochs = 20;
	const int minibatch_size = 10;

	network<sequential> nn;
	nn	<< convolutional_layer(32, 32, 5, 1, 6)  << relu_layer(28, 28, 6)
		<< max_pooling_layer(28, 28, 6, 2)
		<< convolutional_layer(14, 14, 5, 6, 16) << relu_layer(10, 10, 16)
		<< max_pooling_layer(10, 10, 16, 2)
		<< convolutional_layer(5, 5, 5, 16, 120) << relu_layer(1, 1, 120)
		<< fully_connected_layer(120, 10)        << sigmoid_layer(10)
		;

	// load MNIST dataset
	std::vector<label_t> train_labels, test_labels;
	std::vector<vec_t> train_images, test_images;

	std::cout << "load models..." << std::endl;
	parse_mnist_labels(data_dir + "/train-labels-idx1-ubyte", &train_labels);
	parse_mnist_images(data_dir + "/train-images-idx3-ubyte", &train_images, white, black, 2, 2);
	parse_mnist_labels(data_dir + "/t10k-labels-idx1-ubyte", &test_labels);
	parse_mnist_images(data_dir + "/t10k-images-idx3-ubyte", &test_images, white, black, 2, 2);

	adagrad optimizer;
	optimizer.alpha *= (float)std::sqrt(minibatch_size);
	printf("optimizer.alpha = %f\n", optimizer.alpha);

	progress_display disp(train_images.size());
	timer t;
	int epoch_done = 0;
	auto on_enumerate_epoch = [&](){
		printf("%.2fs elapsed. (%d/%d)\n", t.elapsed(), ++epoch_done, max_epochs);

		tiny_dnn::result res = nn.test(test_images, test_labels);
		std::cout << res.num_success << "/" << res.num_total << std::endl;
		disp.restart(train_images.size());
		t.restart();
	};
	auto on_enumerate_minibatch = [&]() { disp += minibatch_size; };
	nn.train<mse>(optimizer, train_images, train_labels, minibatch_size, max_epochs, on_enumerate_minibatch, on_enumerate_epoch);
	std::cout << "end training." << std::endl;
}
/* traning log (new version, simplified):
Threads: <Multi> = <Single>!
epoch 1: 9854/10000
epoch 2: 9884/10000
epoch 3: 9886/10000
epoch 4: 9883/10000
epoch 5: 9885/10000
epoch 6: 9894/10000
epoch 7: 9896/10000
epoch 8: 9903/10000
epoch 9: 9901/10000
epoch 10: 9905/10000
epoch 11: 9908/10000
epoch 12: 9909/10000
epoch 13: 9906/10000
epoch 14: 9906/10000
epoch 15: 9908/10000
epoch 16: 9910/10000
epoch 17: 9909/10000
epoch 18: 9910/10000
epoch 19: 9909/10000
epoch 20: 9908/10000
*/

```

**Comment by beru** [comment]
Can you post git SHA-1 hashes for `tiny-dnn-master-2016-09-22` and `tiny-dnn-master-2017-05-20` please?

**Comment by l0I0I0l** [comment]
For the version `tiny-dnn-master-2016-09-22`:
It is same to use version 1.0.0 alpha1 - the first major version for tiny-dnn
@nyanp nyanp released this Sep 14, 2016
And is the "git SHA-1" the commit `afb8f6d79fed6f8734c6edc532f8dfadfd0e41e7`?

For `tiny-dnn-master-2017-05-20`: just get the current master version.
Any recent version should be same. Can be commit `72fcbfc312c35b96d651c89e170f2493a6deb395`.

**Comment by pliptor** [comment]
How much variation do you see by setting a different initial random seed? I think if there was even a change or reordering of loops somewhere in the algorithms results would not match but hopefully statistically the average would be maintained. Where you using double precision before and now single precision?

**Comment by l0I0I0l** [comment]
It depends on data. Close this issue.
shuf10 = shuffle the training data for 10 rounds.
```
new_shuf10		new_noshuf		old_shuf10		old_noshuf
9858/10000		9854/10000		9868/10000		9877/10000
9883/10000      9884/10000      9890/10000      9901/10000
9893/10000      9886/10000      9897/10000      9912/10000
9905/10000      9883/10000      9897/10000      9918/10000
9904/10000      9885/10000      9902/10000      9921/10000
9905/10000      9894/10000      9910/10000      9922/10000
9910/10000      9896/10000      9912/10000      9919/10000
9907/10000      9903/10000      9915/10000      9917/10000
9910/10000      9901/10000      9915/10000      9921/10000
9911/10000      9905/10000      9913/10000      9923/10000
9915/10000      9908/10000      9911/10000      9924/10000
9911/10000      9909/10000      9912/10000      9925/10000
9914/10000      9906/10000      9909/10000      9924/10000
9913/10000      9906/10000      9909/10000      9927/10000
9913/10000      9908/10000      9908/10000      9924/10000
9917/10000      9910/10000      9908/10000      9923/10000
9916/10000      9909/10000      9908/10000      9923/10000
9916/10000      9910/10000      9908/10000      9924/10000
9914/10000      9909/10000      9908/10000      9924/10000
9913/10000      9908/10000      9910/10000      9925/10000
```
